library/dnet.c

```diff
@@ -393,6 +404,375 @@ static int dnet_iterator_callback_send(void *priv, void *data, uint64_t dsize)
 	return dnet_send_reply_threshold(send->st, send->cmd, data, dsize, 1);
 }
 
+struct dnet_iterator_server_send_write_private {
+	atomic_t			refcnt;
+	struct dnet_server_send_ctl	*send;
+	uint64_t			dsize;
+	char				data[0];
+};
+
+static int dnet_iterator_server_send_complete(struct dnet_addr *addr, struct dnet_cmd *cmd, void *priv)
+{
+	struct dnet_iterator_server_send_write_private *wp = priv;
+	struct dnet_server_send_ctl *send = wp->send;
+	struct dnet_net_state *st = send->state;
+	int err = 0;
+
+	(void) addr;
+
+	/*
+	 * We only care about write transaction completion, thus we only check number
+	 * of transaction destructions. There will be many transactions (one for each remote
+	 * group to write data into) and therefore this code chunk will be invoked multiple times.
+	 *
+	 * If we would like to accumulate write results, we would have to check transaction
+	 * replies and associated data.
+	 */
+	if (is_trans_destroyed(cmd)) {
+		err = -ENOENT;
+		if (cmd)
+			err = cmd->status;
+
+		/*
+		 * Write CAS error is not a real 'error' in that regard, that we do not remove
+		 * local record, but also do not stop iterator.
+		 *
+		 * Setting @write_error forces iterator to stop.
+		 */
+		if (err && !send->write_error && (err != -EBADFD))
+			send->write_error = err;
+
+		if (atomic_dec_and_test(&wp->refcnt)) {
+			if (send->iflags & DNET_IFLAGS_MOVE) {
+				if (!err) {
+					struct dnet_io_req *r;
+					struct dnet_cmd *lc;
+					struct dnet_io_attr *io;
+					size_t cmd_size = sizeof(struct dnet_io_req) +
+							sizeof(struct dnet_cmd) +
+							sizeof(struct dnet_io_attr);
+
+					r = malloc(cmd_size);
+					if (!r) {
+						err = -ENOMEM;
+						// we could update wp->data here, which is dnet_iterator_response
+						// but we do not really care about local errors, for example
+						// remove error is not handled too
+						goto err_out_send;
+					}
+
+					memset(r, 0, cmd_size);
+
+					r->header = r + 1;
+					r->hsize = sizeof(struct dnet_cmd);
+					r->data = r->header + sizeof(struct dnet_cmd);
+					r->dsize = sizeof(struct dnet_io_attr);
+					r->st = dnet_state_get(st);
+
+					lc = r->header;
+					dnet_setup_id(&lc->id, send->cmd.id.group_id, cmd->id.id);
```

Comment:
cmd can not be NULL, check above is being copied from callback to callback. Command structure is either handcrafted in stack when there is error on the client side, or it is stored in transaction, when it is a reply.